scipy/special/tests/test_hyp2f1.py

```diff
@@ -0,0 +1,608 @@
+import pytest
+import numpy as np
+from collections import namedtuple
+from numpy.testing import assert_allclose
+
+from scipy.special import hyp2f1
+from scipy.special._testutils import check_version, MissingModule
+
+
+try:
+    import mpmath
+except ImportError:
+    mpmath = MissingModule("mpmath")
+
+
+Hyp2f1TestCase = namedtuple(
+    "Hyp2f1TestCase",
+    ["a", "b", "c", "z", "expected", "rtol"]
+)
+
+
+class TestHyp2f1:
+    """Tests for hyp2f1 for complex values.
+
+    Expected values for test cases were computed using mpmath. See
+    `scipy.special._precompute.hyp2f1_data`. The verbose style of specifying
+    test cases is used for readability and to make it easier to mark individual
+    cases as expected to fail. Expected failures are used to highlight cases
+    where improvements are needed. See
+    `scipy.special._precompute.hyp2f1_data.make_hyp2f1_test_cases`
+    for a function to generate the boilerplate for the test cases.
+
+    Assertions have been added to each test to ensure that the test cases
+    match the situations that are intended. A final test `test_test_hyp2f1`
+    checks that the expected values in the test cases actually match what
+    is computed by mpmath. This test is marked slow even though it isn't
+    particularly slow so that it won't run on continuous integration builds.
```

Comment:
Shouldn't the tests in this class be skipped if the data file doesn't exist? If that is done somewhere, I'm failing to spot it.

Reply:
Ah, I should have been more clear here. There is actually no data file used. This test is gathering the test cases out of the pytest marks and recomputing the expected value with mpmath to check if it matches. I was little worried that this just a little too clever in the bad sense though. I'm also not sure if the structure of the pytest marks could somehow change down the line, breaking this test, making me feel it's a little brittle.

Reply:
Oh, I understand the cause of the confusion now. The test imports something called `mp_hyp2f1` from `scipy.special._precompute.hyp2f1_data`. It's more reasonable than not to expect this to do something with precomputed values, due to the name of the module, and because one should expect the test just use to `mpmath.hyp2f1` directly. The only source of precomputed values I've given is an enormous file generated by that script.

What's really happening is that `mp_hyp2f1` is a modification of `mpmath.hyp2f1` that returns the same value as `scipy.special.hyp2f1` on the branch cut, where choice of value is discretionary. I placed `mp_hyp2f1` in the file for the precompute script because at first that was the only place it was used. When I later added `test_test_hyp2f1`, I imported it from _precompute, which is pretty confusing now that I think about it.

Is there a better place to put this function? For now I've just added a comment to `test_hyp2f1.py` explaining what the function does.

Reply:
> I'm also not sure if the structure of the pytest marks could somehow change down the line, breaking this test, making me feel it's a little brittle. Edit: I did a quick study of the pytest documentation, and the source code and its git history. It's given me some confidence that this way of getting the parameters out of the test cases should be pretty stable.

I think that's fine then. We have a lot of tests that use this kind of parameter parsing (mostly custom rather than `pytest`-based, because it predates our migration to pytest). And even if it breaks at some point, it'll likely do so loudly so we can just patch it up.

Reply:
> Is there a better place to put this function? For now I've just added a comment to `test_hyp2f1.py` explaining what the function does.

If the comment is clear enough it's fine. Otherwise, you could have turned it around? Put `mp_hyp2f1` into the `test_` file and make the `_precompute` functionality import it from there?

Reply:
I moved it around to put `mp_hyp2f1` in the test and imported it in `_precompute` from there. I was cargo culting some nonsense about not importing from the test directory, but it seems fine to do it here and much better than the alternative.

Reply:
This looks much better, thanks!